### PR TITLE
Enable parallelization for insert+search in HNSW - MOD-4628

### PR DESCRIPTION
**Pull request by alonre24**
**Describe the changes in the pull request**

Enable parallelization in HNSW for every insert and read operation, as long as insertion does override an existing label and resize is not triggered. This includes introducing the fine-grained lock-per-vector mechanism, in which a node's neighbors list is protected by a designated lock before accessing or modifying it.

- Add a new `IN_PROCCESS` flag for every vector to indicate whether the vector is still being inserted to the index (and if so, ignore it upon searching candidates in the "search layer" method).
- Add helper method to HNSW for safe check if a label exists in the index (currently for testing, will be used in the future by the tiered index)
- Create a new unit test file for parallel testing, and move tests into it from HNSW single/multi test files.

### CODE REVIEW

File: src/VecSim/algorithms/hnsw/hnsw.h [hunk 1]
```diff
@@ -1235,133 +1329,134 @@ void HNSWIndex<DataType, DistType>::appendVector(const void *vector_data, const
...
     if (cur_element_count == 0) {
         return entrypoint_node_;
     }
-    idType currObj = entrypoint_node_;
-    DistType cur_dist =
-        this->dist_func(query_data, getDataByInternalId(entrypoint_node_), this->dim);
-    for (size_t level = maxlevel_; level > 0 && currObj != HNSW_INVALID_ID; level--) {
-        greedySearchLevel<true>(query_data, level, currObj, cur_dist, timeoutCtx, rc);
+    idType curr_element = entrypoint_node_;
+    lock.unlock();
```
Comment (GuyAv46) [inline src/VecSim/algorithms/hnsw/hnsw.h:1453]: consider writing a thread-safe `getEntryPoint` function

File: src/VecSim/algorithms/hnsw/hnsw.h [hunk 2]
```diff
@@ -644,116 +663,198 @@ void HNSWIndex<DataType, DistType>::getNeighborsByHeuristic2(
...
+
+    for (idType selected_neighbor : selected_neighbors) {
+        std::unique_lock<std::mutex> node_lock;
+        std::unique_lock<std::mutex> neighbor_lock;
+        idType lower_id = (new_node_id < selected_neighbor) ? new_node_id : selected_neighbor;
+        if (lower_id == new_node_id) {
+            node_lock = std::unique_lock<std::mutex>(element_neighbors_locks_[new_node_id]);
+            neighbor_lock =
+                std::unique_lock<std::mutex>(element_neighbors_locks_[selected_neighbor]);
         } else {
```
Comment (GuyAv46) [inline src/VecSim/algorithms/hnsw/hnsw.h:819]: This should be fenced somehow to prevent compiler or CPU reordering

File: src/VecSim/algorithms/hnsw/hnsw.h [hunk 3]
```diff
@@ -644,116 +663,198 @@ void HNSWIndex<DataType, DistType>::getNeighborsByHeuristic2(
...
-                    neighbour_incoming_edges->push_back(node_id);
-                }
-            }
+        if (cur_node_neighbors_count == max_M_cur) {
+            // The new node cannot add more neighbors
+            continue;
+        }
+        // validations...
+        assert(cur_node_neighbors_count <= max_M_cur && "Neighbors number exceeds limit");
+        assert(selected_neighbor != new_node_id && "Trying to connect an element to itself");
```
Comment (GuyAv46) [inline src/VecSim/algorithms/hnsw/hnsw.h:836]: consider putting this before continuing.

File: src/VecSim/algorithms/hnsw/hnsw.h [hunk 4]
```diff
@@ -644,116 +663,198 @@ void HNSWIndex<DataType, DistType>::getNeighborsByHeuristic2(
...
-                if (it != node_incoming_edges->end()) {
-                    node_incoming_edges->erase(it);
-                } else {
-                    neighbour_incoming_edges->push_back(node_id);
-                }
-            }
+        if (cur_node_neighbors_count == max_M_cur) {
+            // The new node cannot add more neighbors
+            continue;
+        }
```
Comment (GuyAv46) [inline src/VecSim/algorithms/hnsw/hnsw.h:833]: Why `continue` and not `break`?

File: src/VecSim/algorithms/hnsw/hnsw.h [hunk 5]
```diff
@@ -644,116 +663,198 @@ void HNSWIndex<DataType, DistType>::getNeighborsByHeuristic2(
...
-    vecsim_stl::vector<idType> selectedNeighbors(this->allocator);
-    selectedNeighbors.reserve(M_);
-    while (top_candidates.size() > 0) {
-        selectedNeighbors.push_back(top_candidates.top().second);
-        top_candidates.pop();
+    // Collect the existing neighbors and the new node as the neighbor's neighbors candidates.
+    candidatesMaxHeap<DistType> candidates(this->allocator);
+    DistType dist_cur_node_neighbor = this->dist_func(
+        getDataByInternalId(new_node_id), getDataByInternalId(selected_neighbor), this->dim);
+    candidates.emplace(dist_cur_node_neighbor, new_node_id);
```
Comment (GuyAv46) [inline src/VecSim/algorithms/hnsw/hnsw.h:675]: You should have this distance already calculated. consider using it and save recalculation

File: src/VecSim/algorithms/hnsw/hnsw.h [hunk 6]
```diff
@@ -644,116 +663,198 @@ void HNSWIndex<DataType, DistType>::getNeighborsByHeuristic2(
...
-            *ll_cur = *cur_neighbor;
-            ll_cur++;
+    auto *neighbour_incoming_edges = getIncomingEdgesPtr(selected_neighbor, level);
+    size_t neighbor_neighbors_count = getListCount(neighbor_neighbors_list);
+
+    size_t neighbour_neighbours_idx = 0;
+    bool update_cur_node_required = true;
+    for (size_t i = 0; i < neighbor_neighbors_count; i++) {
+        if (!std::binary_search(nodes_to_update.begin(), nodes_to_update.end(),
+                                neighbor_neighbors_list[i])) {
```
Comment (GuyAv46) [inline src/VecSim/algorithms/hnsw/hnsw.h:732]: is this tested to be faster than a regular search?

File: src/VecSim/algorithms/hnsw/hnsw.h [hunk 7]
```diff
@@ -644,116 +663,198 @@ void HNSWIndex<DataType, DistType>::getNeighborsByHeuristic2(
...
+            // the neighbor is not in the "to_update" nodes list - leave it as is.
+            neighbor_neighbors_list[neighbour_neighbours_idx++] = neighbor_neighbors_list[i];
+            continue;
+        } else if (neighbor_neighbors_list[i] == new_node_id) {
+            // the new node is somehow got into the neighbor's neighbours in the meantime - leave it
```
Comment (GuyAv46) [inline src/VecSim/algorithms/hnsw/hnsw.h:737]: "somehow" meaning while we unlocked and locked again?

File: src/VecSim/algorithms/hnsw/hnsw.h [hunk 8]
```diff
@@ -644,116 +663,198 @@ void HNSWIndex<DataType, DistType>::getNeighborsByHeuristic2(
...
+            removed_node_incoming_edges->erase(it);
+        } else {
+            neighbour_incoming_edges->push_back(removed_node);
         }
+    }
 
-        auto *incoming_edges = new (this->allocator) vecsim_stl::vector<idType>(this->allocator);
-        setIncomingEdgesPtr(cur_c, level, (void *)incoming_edges);
+    size_t cur_node_neighbors_count = getListCount(new_node_neighbors_list);
+    if (update_cur_node_required && cur_node_neighbors_count < max_M_cur &&
```
Comment (GuyAv46) [inline src/VecSim/algorithms/hnsw/hnsw.h:764]: Why do we need to enforce this update?

**Review by GuyAv46 — COMMENTED** [review]
Very very nice!
Few notes

File: tests/unit/test_hnsw_parallel.cpp [hunk 9]
```diff
@@ -0,0 +1,405 @@
...
+                         .initialCapacity = n,
+                         .M = 16,
+                         .efConstruction = 200};
+
+    VecSimIndex *parallel_index = this->CreateNewIndex(params, true);
+    size_t n_threads = 10;
+
+    auto parallel_insert = [&](int myID) {
+        for (size_t i = myID; i < n; i += n_threads) {
+            GenerateAndAddVector<TEST_DATA_T>(parallel_index, dim, i % n_labels, i);
```
[the remark this answers is not in the document]
Comment (alonre24) [reply tests/unit/test_hnsw_parallel.cpp:307]: What do you mean? did you confuse `n_threads` with `n_labels` maybe?

File: src/VecSim/algorithms/hnsw/hnsw.h [hunk 10]
```diff
@@ -147,9 +145,13 @@ class HNSWIndex : public VecSimIndexAbstract<DistType>,
...
     void getNeighborsByHeuristic2(candidatesMaxHeap<DistType> &top_candidates, size_t M);
-    inline idType mutuallyConnectNewElement(idType cur_c,
-                                            candidatesMaxHeap<DistType> &top_candidates,
-                                            size_t level);
+    void revisitNeighborConnections(size_t level, idType new_node_id, idType selected_neighbor,
+                                    idType *new_node_neighbors_list,
+                                    idType *neighbor_neighbors_list,
+                                    std::unique_lock<std::mutex> &node_lock,
+                                    std::unique_lock<std::mutex> &neighbor_lock);
+    idType mutuallyConnectNewElement(idType new_node_id,
```
[the remark this answers is not in the document]
Comment (alonre24) [reply src/VecSim/algorithms/hnsw/hnsw.h:153]: yes

File: src/VecSim/algorithms/hnsw/hnsw.h [hunk 11]
```diff
@@ -95,11 +95,9 @@ class HNSWIndex : public VecSimIndexAbstract<DistType>,
...
     // This is mutable since the object changes upon search operations as well (which are const).
     mutable VisitedNodesHandlerPool visited_nodes_handler_pool;
-#ifdef ENABLE_PARALLELIZATION
-    std::mutex global;
-    std::mutex cur_element_count_guard_;
-    std::vector<std::mutex> link_list_locks_;
-#endif
+    mutable std::mutex entry_point_guard_;
+    mutable std::mutex index_data_guard_;
+    mutable vecsim_stl::vector<std::mutex> element_neighbors_locks_;
```
[the remark this answers is not in the document]
Comment (alonre24) [reply src/VecSim/algorithms/hnsw/hnsw.h:100]: Because they are modified in `const` functions (search methods)

File: src/VecSim/algorithms/hnsw/hnsw.h [hunk 12]
```diff
@@ -1235,133 +1329,134 @@ void HNSWIndex<DataType, DistType>::appendVector(const void *vector_data, const
...
     // this condition only means that we are not inserting the first element.
-    if (entrypoint_node_ != HNSW_INVALID_ID) {
+    if (curr_element != HNSW_INVALID_ID) {
         DistType cur_dist = std::numeric_limits<DistType>::max();
-        if (element_max_level < maxlevelcopy) {
-            cur_dist = this->dist_func(vector_data, getDataByInternalId(currObj), this->dim);
-            for (size_t level = maxlevelcopy; level > element_max_level; level--) {
+        if (element_max_level < max_level_copy) {
+            cur_dist = this->dist_func(vector_data, getDataByInternalId(curr_element), this->dim);
+            for (int level = max_level_copy; level > element_max_level; level--) {
```
[the remark this answers is not in the document]
Comment (alonre24) [reply src/VecSim/algorithms/hnsw/hnsw.h:1374]: if the index is empty, `max_level_copy` is -1

File: src/VecSim/algorithms/hnsw/hnsw.h [hunk 13]
```diff
@@ -1235,133 +1329,134 @@ void HNSWIndex<DataType, DistType>::appendVector(const void *vector_data, const
...
 
         // updating the maximum level (holding a global lock)
-        if (element_max_level > maxlevelcopy) {
-            entrypoint_node_ = cur_c;
+        if (element_max_level > max_level_copy) {
+            entrypoint_node_ = new_element_id;
             maxlevel_ = element_max_level;
             // create the incoming edges set for the new levels.
-            for (size_t level_idx = maxlevelcopy + 1; level_idx <= element_max_level; level_idx++) {
+            for (int level_idx = max_level_copy + 1; level_idx <= element_max_level; level_idx++) {
```
[the remark this answers is not in the document]
Comment (alonre24) [reply src/VecSim/algorithms/hnsw/hnsw.h:1419]: if the index is empty, max_level_copy is -1. We're not suppose to get there if that the case, but it's safer to use int...

File: src/VecSim/algorithms/hnsw/hnsw.h (hunk 2 above)
Comment (alonre24) [reply src/VecSim/algorithms/hnsw/hnsw.h:819]: intersting. I'll look into that

File: src/VecSim/algorithms/hnsw/hnsw.h (hunk 6 above)
Comment (alonre24) [reply src/VecSim/algorithms/hnsw/hnsw.h:732]: What do you mean by regular search? Since we know that `nodes_to_update` is already sorted, using binary search is the trivially better option no?

File: src/VecSim/algorithms/hnsw/hnsw.h (hunk 7 above)
Comment (alonre24) [reply src/VecSim/algorithms/hnsw/hnsw.h:737]: yes :sweat_smile:
I'll rephrase the comment

File: src/VecSim/algorithms/hnsw/hnsw.h (hunk 2 above)
Comment (alonre24) [reply src/VecSim/algorithms/hnsw/hnsw.h:819]: After checking with the experts, it's not required:) The compiler assures the fence between both locks implicitly

File: src/VecSim/algorithms/hnsw/hnsw.h (hunk 4 above)
Comment (alonre24) [reply src/VecSim/algorithms/hnsw/hnsw.h:833]: right. Even in the pseudo-code in the doc it says break :)

File: src/VecSim/algorithms/hnsw/hnsw.h (hunk 8 above)
Comment (alonre24) [reply src/VecSim/algorithms/hnsw/hnsw.h:764]: What do you mean?

File: src/VecSim/algorithms/hnsw/hnsw.h [hunk 14]
```diff
@@ -1045,13 +1145,8 @@ HNSWIndex<DataType, DistType>::HNSWIndex(const HNSWParams *params,
...
       data_size_(VecSimType_sizeof(params->type) * this->dim),
       element_levels_(max_elements_, allocator),
-      visited_nodes_handler_pool(pool_initial_size, max_elements_, this->allocator)
-
-#ifdef ENABLE_PARALLELIZATION
-      ,
-      link_list_locks_(max_elements_)
-#endif
-{
+      visited_nodes_handler_pool((int)pool_initial_size, max_elements_, allocator),
```
[the remark this answers is not in the document]
Comment (alonre24) [reply src/VecSim/algorithms/hnsw/hnsw.h]: no reason

File: tests/unit/test_hnsw_parallel.cpp [hunk 15]
```diff
@@ -0,0 +1,405 @@
...
+    size_t memory_before = index->info().hnswInfo.memory;
+    size_t n_threads = 16;
+    std::thread thread_objs[n_threads];
+    for (size_t i = 0; i < n_threads; i++) {
+        thread_objs[i] = std::thread(parallel_search, i);
+    }
+    for (size_t i = 0; i < n_threads; i++) {
+        thread_objs[i].join();
+    }
+    ASSERT_EQ(successful_searches, n_threads);
```
[the remark this answers is not in the document]
Comment (alonre24) [reply tests/unit/test_hnsw_parallel.cpp:86]: Add debug info fields for checking the parallel workers. It is flaky, but it is something

File: src/VecSim/algorithms/hnsw/hnsw.h [hunk 16]
```diff
@@ -1437,23 +1464,29 @@ void HNSWIndex<DataType, DistType>::appendVector(const void *vector_data, const
...
-        return entrypoint_node_;
-    }
-    idType curr_element = entrypoint_node_;
-    lock.unlock();
+    idType curr_element = safeGetEntryPointCopy();
+    if (curr_element == HNSW_INVALID_ID)
+        return curr_element; // index is empty.
 
     DistType cur_dist = this->dist_func(query_data, getDataByInternalId(curr_element), this->dim);
-    for (size_t level = maxlevel_; level > 0 && curr_element != HNSW_INVALID_ID; level--) {
```
Comment (GuyAv46) [inline src/VecSim/algorithms/hnsw/hnsw.h]: Why did you remove this condition? `curr_element` can become `HNSW_INVALID_ID` if a timeout occurred in `greedySearchLevel<true>`

File: src/VecSim/algorithms/hnsw/hnsw_batch_iterator.h [hunk 17]
```diff
@@ -78,7 +83,7 @@ HNSW_BatchIterator<DataType, DistType>::HNSW_BatchIterator(
 
     this->dist_func = index->getDistFunc();
     this->dim = index->getDim();
-    this->entry_point = index->getEntryPointId();
+    this->entry_point = index->safeGetEntryPointCopy();
```
Comment (GuyAv46) [inline src/VecSim/algorithms/hnsw/hnsw_batch_iterator.h:86]: Can entering from an out-of-date entry point can cause issues? maybe we should call `safeGetEntryPointCopy` just before using its value?

**Review by GuyAv46 — APPROVED** [review]

File: .github/workflows/benchmark.yml [hunk 18]
```diff
@@ -75,7 +75,8 @@ jobs:
           ./tests/benchmark/benchmarks.sh ${{ github.event.label.name }} | xargs -I {} redisbench-admin export     \
           --redistimeseries_host      ${{ secrets.PERFORMANCE_RTS_HOST }}           \
           --redistimeseries_port      ${{ secrets.PERFORMANCE_RTS_PORT }}           \
-          --redistimeseries_pass      ${{ secrets.PERFORMANCE_RTS_AUTH }}           \
+          --redistimeseries_user      default                                       \
+          --redistimeseries_pass      '${{ secrets.PERFORMANCE_RTS_AUTH }}'         \
```
[the remark this answers is not in the document]
Comment (alonre24) [reply .github/workflows/benchmark.yml:79]: I did. Don't know why it still appears in the diff

File: src/VecSim/vec_sim_common.h [hunk 19]
```diff
@@ -34,7 +34,10 @@ typedef enum { VecSimAlgo_BF, VecSimAlgo_HNSWLIB } VecSimAlgo;
 typedef enum { VecSimMetric_L2, VecSimMetric_IP, VecSimMetric_Cosine } VecSimMetric;
 
 // Vectors flags (for marking a specific vector)
-typedef enum { DELETE_MARK = 0x01 } Flags;
+typedef enum {
+    DELETE_MARK = 0x01, // vector is logically deleted, but still exists in the graph
+    IN_PROCESS = 0x02   // vector is being inserted into the graph
```
[the remark this answers is not in the document]
Comment (alonre24) [reply src/VecSim/vec_sim_common.h:39]: If not here then where? in `hnsw.h`?

File: src/VecSim/utils/vec_utils.h [hunk 20]
```diff
@@ -81,7 +81,7 @@ void normalizeVector(DataType *input_vector, size_t dim) {
     double sum = 0;
 
     for (size_t i = 0; i < dim; i++) {
-        sum += input_vector[i] * input_vector[i];
+        sum += (double)input_vector[i] * (double)input_vector[i];
```
[the remark this answers is not in the document]
Comment (alonre24) [reply src/VecSim/utils/vec_utils.h:84]: I fixed it due to a security alert...

File: src/VecSim/algorithms/hnsw/hnsw.h [hunk 21]
```diff
@@ -644,116 +680,202 @@ void HNSWIndex<DataType, DistType>::getNeighborsByHeuristic2(
...
+    candidates.emplace(neighbor_data.first, new_node_id);
+
+    idType selected_neighbor = neighbor_data.second;
+    for (size_t j = 0; j < getListCount(neighbor_neighbors_list); j++) {
+        candidates.emplace(this->dist_func(getDataByInternalId(neighbor_neighbors_list[j]),
+                                           getDataByInternalId(selected_neighbor), this->dim),
+                           neighbor_neighbors_list[j]);
+    }
+
+    std::vector<idType> nodes_to_update;
```
[the remark this answers is not in the document]
Comment (alonre24) [reply src/VecSim/algorithms/hnsw/hnsw.h:700]: not exactly, because we add the new node id and the neighbor to this collection as well...

File: src/VecSim/algorithms/hnsw/hnsw.h [hunk 22]
```diff
@@ -1045,13 +1166,8 @@ HNSWIndex<DataType, DistType>::HNSWIndex(const HNSWParams *params,
...
       element_levels_(max_elements_, allocator),
-      visited_nodes_handler_pool(pool_initial_size, max_elements_, this->allocator)
-
-#ifdef ENABLE_PARALLELIZATION
-      ,
-      link_list_locks_(max_elements_)
-#endif
-{
+      visited_nodes_handler_pool(pool_initial_size, max_elements_, allocator),
+      element_neighbors_locks_(max_elements_, allocator) {
```
[the remark this answers is not in the document]
Comment (alonre24) [reply src/VecSim/algorithms/hnsw/hnsw.h:1170]: discussed orally..

File: tests/unit/test_hnsw_parallel.cpp [hunk 23]
```diff
@@ -286,7 +307,10 @@ TYPED_TEST(HNSWTestParallel, parallelInsert) {
         thread_objs[i].join();
     }
     ASSERT_EQ(VecSimIndex_IndexSize(parallel_index), n);
-    ASSERT_EQ(this->CastToHNSW(parallel_index)->max_parallel_workers, n_threads);
+    // Validate that every thread executed a single job.
```
Comment (DvirDukhan) [inline tests/unit/test_hnsw_parallel.cpp:310]:
```suggestion
    // Validate that every thread executed `n / n_threads` jobs.
```

File: tests/unit/test_hnsw_parallel.cpp [hunk 24]
```diff
@@ -330,7 +358,10 @@ TYPED_TEST(HNSWTestParallel, parallelInsertMulti) {
...
```
Comment (DvirDukhan) [inline tests/unit/test_hnsw_parallel.cpp:361]: same here

File: tests/unit/test_hnsw_parallel.cpp [hunk 25]
```diff
@@ -415,6 +451,17 @@ TYPED_TEST(HNSWTestParallel, parallelInsertSearch) {
         }
         ASSERT_EQ(VecSimIndex_IndexSize(parallel_index), n);
         ASSERT_EQ(successful_searches, ceil(double(n_threads) / 2));
-        ASSERT_GT(this->CastToHNSW(parallel_index)->max_parallel_workers, n_threads / 2);
+        // Validate that every thread executed a single job.
```
Comment (DvirDukhan) [inline tests/unit/test_hnsw_parallel.cpp:454]: same here, fix the comment please

**Review by DvirDukhan — APPROVED** [review]
